File: 2-AddTwoNumbers/2-AddTwoNumbers.java
// Last updated: 4/29/2026, 10:37:03 AM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode node1 = l1, node2 = l2;
        ListNode res = null, node3 = null;
        int carry = 0;
        while(node1 != null || node2 != null) {
            int val1 = 0, val2 = 0;
            if (node1 != null) {
                val1 = node1.val;    
                node1 = node1.next;
            }
            if (node2 != null) {
                val2 = node2.val;
                node2 = node2.next;
            }
            int val3 = val1 + val2 + carry;
            carry = val3 / 10;
            val3 = val3 % 10;
            
            if(node3 == null) {
                node3 = new ListNode(val3);
                res = node3;
            } else {
                node3.next = new ListNode(val3);
                node3 = node3.next;
            }
        }
        
        if (carry != 0) {
            node3.next = new ListNode(carry);
        }
        
        return res;
    }
}

File: 2-AddTwoNumbers/2-AddTwoNumbers.js
// Last updated: 4/29/2026, 10:37:00 AM
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let prevOut;
    let root;
    let curr1 = l1;
    let curr2 = l2;
    let carry = 0;
    while(curr1 || curr2) {
        const val1 = curr1 !== null? curr1.val: 0;
        const val2 = curr2 !== null? curr2.val: 0;
        const sum = val1 + val2 + carry;
        const remainder = sum % 10;
        carry = Math.floor(sum / 10);
        if(!prevOut) {
            root = new ListNode(remainder, null);
            prevOut = root;
        } else {
            prevOut.next = new ListNode(remainder, null);
            prevOut = prevOut.next;
        }
        if(curr1) {
            curr1 = curr1.next;
        }
        if(curr2) {
            curr2 = curr2.next;
        }
    }
    if(carry > 0) {
        prevOut.next = new ListNode(carry, null);
    }
    return root;
};